File: week-3/PermutationTransformation.py

class Solution:
    def update(self, ls, arr, level):
        if len(ls) == 0:
            return
        max_size = max(ls)
        max_index = ls.index(max_size)
        arr[max_size] = level
        self.update(ls[:max_index], arr, level+1)
        self.update(ls[max_index+1:], arr, level+1)
        return

    def permutationTransformation(self, n, ls):
        arr = {}
        self.update(ls, arr, 0)
        return arr


if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        n = int(input())
        ls = list(map(int, input().split()))
        s = Solution()
        arr = s.permutationTransformation(n, ls)
        for i in range(n):
            print(arr[ls[i]], end=' ')
        print()
        

File: week-3/PermutationTransformation.cpp
#include <bits/stdc++.h>
using namespace std;

void update(vector<int> ls, map<int, int> &arr, int level)
{
    if (ls.size() == 0)
    {
        return;
    }
    int max = *max_element(ls.begin(), ls.end());
    arr[max] = level;
    int max_index = distance(ls.begin(), find(ls.begin(), ls.end(), max));
    update(vector<int>(ls.begin(), ls.begin() + max_index), arr, level + 1);
    update(vector<int>(ls.begin() + max_index + 1, ls.end()), arr, level + 1);
    return;
}

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n;
        cin >> n;
        vector<int> ls;
        for (int i = 0; i < n; i++)
        {
            int temp;
            cin >> temp;
            ls.push_back(temp);
        }
        map<int, int> arr;
        update(ls, arr, 0);
        for (int i = 0; i < n; i++)
        {
            cout << arr[ls[i]] << " ";
        }
        cout << endl;
    }
    return 0;
}